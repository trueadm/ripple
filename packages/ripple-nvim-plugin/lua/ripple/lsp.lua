local M = {}

local SERVER_NAME = "ripple"
local LSP_PACKAGE = "ripple-language-server"
local EXACT_VERSION_PATTERN = "^%d+%.%d+%.%d+$"

local function plugin_package_json_path()
	local plugin_root = debug.getinfo(1, "S").source:sub(2)
	return vim.fs.find("package.json", { upward = true, path = vim.fs.dirname(plugin_root) })[1]
end

local function read_json(path)
	local stat = vim.loop.fs_stat(path)
	if not stat or stat.type ~= "file" then
		return nil
	end

	local ok, contents = pcall(vim.fn.readfile, path)
	if not ok or type(contents) ~= "table" then
		return nil
	end

	local json_string = table.concat(contents, "")

	local decoded
	local ok_decode, value = pcall(vim.json.decode, json_string)
	if ok_decode then
		decoded = value
	end

	if type(decoded) ~= "table" then
		return nil
	end

	return decoded
end

local function resolve_required_version()
	local package_json_path = plugin_package_json_path()

	local package_json = read_json(package_json_path)
	if not package_json then
		return nil, "unable to read plugin package.json"
	end

	local config = package_json.config or {}

	local spec = config["ripple-language-server"]

	if type(spec) ~= "string" or spec == "" then
		return nil, "missing config.rippleLanguageServerVersion field in package.json"
	end

	spec = spec:gsub("^%s+", ""):gsub("%s+$", "")

	if not spec:match(EXACT_VERSION_PATTERN) then
		return nil, ("unsupported version spec '%s'; please pin an exact semver"):format(spec)
	end

	return spec, nil
end

local function installed_version(install_dir)
	local pkg = read_json(install_dir .. "/node_modules/" .. LSP_PACKAGE .. "/package.json")
	if not pkg or type(pkg.version) ~= "string" then
		return nil
	end
	return pkg.version
end

local function ensure_server_binary()
	local required_version, err = resolve_required_version()
	if not required_version then
		vim.notify("[ripple] " .. err, vim.log.levels.ERROR)
		return nil
	end

	local install_dir = vim.fn.stdpath("data") .. "/" .. LSP_PACKAGE
	local bin = install_dir .. "/node_modules/.bin/" .. LSP_PACKAGE

	if vim.loop.os_uname().version:match("Windows") then
		bin = bin .. ".cmd"
	end

	local function version_matches()
		local version = installed_version(install_dir)
		return version == required_version
	end

	if vim.fn.executable(bin) == 1 and version_matches() then
		return bin
	end

	vim.fn.mkdir(install_dir, "p")
	vim.notify(("[ripple] Installing %s@%s..."):format(LSP_PACKAGE, required_version), vim.log.levels.INFO)

	local out = vim.fn.system({
		"npm",
		"install",
		("%s@%s"):format(LSP_PACKAGE, required_version),
		"--prefix",
		install_dir,
		"--no-audit",
		"--no-fund",
	})

	if vim.v.shell_error ~= 0 then
		vim.notify("[ripple] npm install failed:\n" .. out, vim.log.levels.ERROR)
		return nil
	end

	if not version_matches() then
		local found = installed_version(install_dir) or "unknown"
		vim.notify(
			("[ripple] Installed %s but required %s"):format(found, required_version),
			vim.log.levels.ERROR
		)
		return nil
	end

	return bin
end

function M.setup()
	local bin = ensure_server_binary()
	if not bin then
		return
	end

	local base_config = {
		cmd = { bin, "--stdio" },
		filetypes = { "ripple" },
		root_markers = { "package.json", "pnpm-workspace.yaml", ".git" },
	}

	vim.lsp.config(SERVER_NAME, base_config)
	vim.lsp.enable(SERVER_NAME)
end

return M
