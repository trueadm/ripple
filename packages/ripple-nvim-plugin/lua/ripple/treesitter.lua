local M = {}

function M.setup()
	local ok, parsers = pcall(require, "nvim-treesitter.parsers")
	if not ok then
		vim.schedule(function()
			vim.notify("[ripple] nvim-treesitter not found; skipping parser registration", vim.log.levels.WARN)
		end)
		return
	end

	local parser_config = parsers.get_parser_configs()
	local default_install = {
		url = "https://github.com/trueadm/ripple",
		files = {
			"src/parser.c",
			"src/scanner.c",
		},
		branch = "main",
		location = "packages/tree-sitter-ripple",
	}

	local config = parser_config.ripple or {}
	local install_info = config.install_info or {}
	config.install_info = vim.tbl_deep_extend("force", default_install, install_info)
	config.filetype = config.filetype or "ripple"
	config.used_by = config.used_by or { "ripple" }
	parser_config.ripple = config

	local ok_install, install = pcall(require, "nvim-treesitter.install")
	if ok_install and type(install.ensure_installed) == "function" then
		install.ensure_installed("ripple")
	else
		vim.schedule(function()
			vim.notify("[ripple] Install nvim-treesitter parser manually via :TSInstall ripple", vim.log.levels.INFO)
		end)
	end
end

return M
