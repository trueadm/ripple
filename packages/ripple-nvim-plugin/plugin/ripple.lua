if vim.filetype and vim.filetype.add then
	vim.filetype.add({
		extension = {
			ripple = "ripple",
		},
	})
end
