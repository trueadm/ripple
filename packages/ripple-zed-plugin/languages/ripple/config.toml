name = "Ripple"
grammar = "ripple"
path_suffixes = ["ripple"]
line_comments = ["//"]
block_comment = ["/*", "*/"]
tab_size = 2
hard_tabs = false

autoclose_before = ";:.,=}])\""

brackets = [
  { start = "{", end = "}", close = true, newline = true },
  { start = "[", end = "]", close = true, newline = true },
  { start = "(", end = ")", close = true, newline = true },
  { start = "<", end = ">", close = true, newline = false },
  { start = "\"", end = "\"", close = true, newline = false },
  { start = "'", end = "'", close = true, newline = false },
  { start = "`", end = "`", close = true, newline = false },
]
