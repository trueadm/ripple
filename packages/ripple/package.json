{
  "name": "ripple",
  "description": "Ripple is an elegant TypeScript UI framework",
  "license": "MIT",
  "author": "Dominic Gannaway",
  "version": "0.2.67",
  "type": "module",
  "module": "src/runtime/index.js",
  "main": "src/runtime/index.js",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/trueadm/ripple.git",
    "directory": "packages/ripple"
  },
  "bugs": {
    "url": "https://github.com/trueadm/ripple/issues"
  },
  "homepage": "https://ripplejs.com",
  "keywords": [
    "ripple",
    "UI",
    "framework"
  ],
  "exports": {
    ".": {
      "types": "./types/index.d.ts",
      "worker": "./src/runtime/index.js",
      "browser": "./src/runtime/index.js",
      "default": "./src/runtime/index.js"
    },
    "./package.json": "./package.json",
    "./compiler": {
      "types": "./types/index.d.ts",
      "require": "./compiler/index.js",
      "default": "./src/compiler/index.js"
    },
    "./validator": {
      "types": "./types/index.d.ts",
      "require": "./validator/index.js",
      "default": "./src/validator/index.js"
    },
    "./internal/client": {
      "default": "./src/runtime/internal/client/index.js"
    },
    "./jsx-runtime": {
      "types": "./src/jsx-runtime.d.ts",
      "import": "./src/jsx-runtime.js",
      "default": "./src/jsx-runtime.js"
    }
  },
  "imports": {
    "#client": "./src/runtime/internal/client/types.d.ts",
    "#client/constants": "./src/internal/client/constants.js"
  },
  "dependencies": {
    "@jridgewell/sourcemap-codec": "^1.5.5",
    "acorn": "^8.15.0",
    "acorn-typescript": "^1.4.13",
    "esrap": "^2.1.0",
    "is-reference": "^3.0.3",
    "magic-string": "^0.30.18",
    "muggle-string": "^0.4.1",
    "zimmerframe": "^1.1.2"
  },
  "devDependencies": {
    "@types/estree": "^1.0.8",
    "@types/node": "^24.3.0",
    "typescript": "catalog:"
  }
}
