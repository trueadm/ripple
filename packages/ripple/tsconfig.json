{
	"compilerOptions": {
		"module": "esnext",
		"lib": ["esnext", "dom", "dom.iterable"],
		"target": "esnext",
		"noEmit": true,
		"moduleResolution": "bundler",
		"resolveJsonModule": true,
		"noEmitOnError": true,
		"noErrorTruncation": true,
		"allowSyntheticDefaultImports": true,
		"verbatimModuleSyntax": true,
		"types": ["node"],
		"jsx": "preserve",
		"jsxImportSource": "ripple",
		"strict": true,
		"allowJs": true,
		"checkJs": true,
		"paths": {
			"ripple": ["./types/index.d.ts"],
		}
	},
	"include": [
		"./*.js",
		"./src/",
		"./tests/*/test.ripple",
	]
}
