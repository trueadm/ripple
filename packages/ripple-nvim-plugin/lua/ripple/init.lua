local M = {}

---Entrypoint invoked by plugin managers.
---@param opts? {treesitter?: table|boolean, lsp?: table|boolean}
function M.setup(opts)
	opts = opts or {treesitter = { ensure_installed = true }, lsp = {}}

	if opts.treesitter ~= false then
		require("ripple.treesitter").setup(opts.treesitter or {})
	end

	if opts.lsp ~= false then
		require("ripple.lsp").setup(opts.lsp or {})
	end
end

return M
