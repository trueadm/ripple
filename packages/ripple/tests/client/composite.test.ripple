import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { mount, flushSync, track, TrackedArray, TrackedMap, effect } from 'ripple';

describe('composite components', () => {
	let container;

	function render(component) {
		mount(component, {
			target: container
		});
	}

	beforeEach(() => {
		container = document.createElement('div');
		document.body.appendChild(container);
	});

	afterEach(() => {
		document.body.removeChild(container);
		container = null;
	});

	it('renders composite components', () => {
		component Button({ count }) {
			<div>{count}</div>
		}

		component App() {
			let count = track(0);

			<button onClick={() => @count++}>{'Increment'}</button>
			<Button {@count} />
		}

		render(App);

		const button = container.querySelector('button');

		button.click();
		flushSync();

		expect(container.querySelector('div').textContent).toBe('1');

		button.click();
		flushSync();

		expect(container.querySelector('div').textContent).toBe('2');
	});

	it('renders composite components with object state', () => {
		component Button({ obj }) {
			<button class='count2' onClick={() => {
				obj.@count++;
			}}>{obj.@count}</button>
		}

		component App() {
			<div>
				let obj = {
					count: track(0)
				};

				<span class='count'>{obj.@count}</span>
				<Button obj={obj} />
			</div>
		}

		render(App);

		const button = container.querySelector('button');

		button.click();
		flushSync();

		expect(container.querySelector('.count').textContent).toBe('1');
		expect(container.querySelector('.count2').textContent).toBe('1');
	});

	it('renders composite components with object state wrapped in an if statement', () => {
		component Button({ obj }) {
			<button class='count2' onClick={() => {
				obj.@count++;
			}}>{obj.@count}</button>
		}

		component OtherComponent({ obj }) {
			<div class='count3'>{obj.@count}</div>
		}

		component App() {
			<div>
				let obj = {
					count: track(0)
				};

				<span class='count'>{obj.@count}</span>
				<span>{' '}</span>
				if (obj) {
					<Button obj={obj} />
				}

				if (obj) {
					<OtherComponent obj={obj} />
				}
			</div>
		}

		render(App);

		const button = container.querySelector('button');

		button.click();
		flushSync();

		expect(container.querySelector('.count').textContent).toBe('1');
		expect(container.querySelector('.count2').textContent).toBe('1');
		expect(container.querySelector('.count3').textContent).toBe('1');
	});

	it('parents and children have isolated state', () => {
		component Button(props) {
			let count = track(() => props.count);
			<button onClick={() => { @count++; } }>{"child: " + @count}</button>
		}

		component App() {
			<div>
				let count = track(0);

				<button onClick={() => { @count++; } }>{"parent: " + @count}</button>
				<Button {@count} />
			</div>
		}

		render(App);

		const buttons = container.querySelectorAll('button');

		expect(buttons[0].textContent).toBe('parent: 0');
		expect(buttons[1].textContent).toBe('child: 0');

		buttons[0].click();
		flushSync();

		expect(buttons[0].textContent).toBe('parent: 1');
		expect(buttons[1].textContent).toBe('child: 1');

		buttons[1].click();
		flushSync();

		expect(buttons[0].textContent).toBe('parent: 1');
		expect(buttons[1].textContent).toBe('child: 2');
	});

	it('parents and children have isolated connected state (destructured props)', () => {
		component Button({count}) {
			let local_count = track(() => count);
			<button onClick={() => { @local_count++; } }>{"child: " + @local_count}</button>
		}

		component App() {
			<div>
				let count = track(0);

				<button onClick={() => { @count++; } }>{"parent: " + @count}</button>
				<Button {@count} />
			</div>
		}

		render(App);

		const buttons = container.querySelectorAll('button');

		expect(buttons[0].textContent).toBe('parent: 0');
		expect(buttons[1].textContent).toBe('child: 0');

		buttons[0].click();
		flushSync();

		expect(buttons[0].textContent).toBe('parent: 1');
		expect(buttons[1].textContent).toBe('child: 1');

		buttons[1].click();
		flushSync();

		expect(buttons[0].textContent).toBe('parent: 1');
		expect(buttons[1].textContent).toBe('child: 2');
	});

	it('correct handles passing through component props and children', () => {
		component Button({ A, B, children }) {
			<div>
				<A />
				<children />
				<B />
			</div>
		}

		component App() {
			<Button>
				component A() {
					<div>{"I am A"}</div>
				}
				<div>{"other text"}</div>
				component B() {
					<div>{"I am B"}</div>
				}
			</Button>
		}

		render(App);

		expect(container).toMatchSnapshot();
	});

	it('correctly handles default prop values', () => {
		component Child({ foo = 456 }) {
			<div>{foo}</div>
		}

		component App(props) {
			let foo = track(123);

			<Child />
			<Child {@foo} />
		}

		render(App);

		expect(container.querySelectorAll('div')[0].textContent).toBe('456');
		expect(container.querySelectorAll('div')[1].textContent).toBe('123');
	});

	it('correctly handles default prop values #2', () => {
		component Child({ foo = 456 }) {
			<div>{foo}</div>
		}

		component App(props) {
			let foo = 123;

			<Child />
			<Child {foo} />
		}

		render(App);

		expect(container.querySelectorAll('div')[0].textContent).toBe('456');
		expect(container.querySelectorAll('div')[1].textContent).toBe('123');
	});

	it('correctly handles no props', () => {
		component Child(props) {
			<div>{props.@foo}</div>
		}

		component App(props) {
			let foo = track(123);

			<Child />
			<Child {foo} />
		}

		render(App);

		expect(container.querySelectorAll('div')[0].textContent).toBe('');
		expect(container.querySelectorAll('div')[1].textContent).toBe('123');
	});

	it('correctly handles no props #2', () => {
		component Child({ foo }) {
			<div>{foo}</div>
		}

		component App(props) {
			let foo = track(123);

			<Child />
			<Child {@foo} />
		}

		render(App);

		expect(container.querySelectorAll('div')[0].textContent).toBe('');
		expect(container.querySelectorAll('div')[1].textContent).toBe('123');
	});

	it('handlers generic', () => {
		component ArrayTest() {
			let items = new TrackedArray<number>();
			items.push.apply(items, [1, 2, 3, 4, 5]);

			<pre>{items ? JSON.stringify(items) : 'Loading...'}</pre>
		}

		render(ArrayTest);
	});

	it('handles spreading of props', () => {
		let logs = [];

		component App() {
			const a = track(1);
			const b = track(2);
			const c = track(3);

			const obj = track(() => ({
				@a,
				@b,
				@c,
			}));

			<Child {...@obj} />

			<button onClick={() => { @a++; @b++; @c++; }}>{"Increment all"}</button>
		}

		component Child({ a, b, c }) {
			effect(() => {
				logs.push(`Child effect: ${a}, ${b}, ${c}`);
			});

			<div>{a + ' ' +  b + ' ' + c}</div>
		}

		render(App);
		flushSync();

		expect(container.querySelector('div').textContent).toBe('1 2 3');
		expect(logs).toEqual(['Child effect: 1, 2, 3']);

		const button = container.querySelector('button');
		button.click();
		flushSync();

		expect(container.querySelector('div').textContent).toBe('2 3 4');
		expect(logs).toEqual(['Child effect: 1, 2, 3', 'Child effect: 2, 3, 4']);
	});

	it('handles advanced generic ambiguity and edge cases', () => {
		component App() {
			// Ambiguous generics vs JSX / less-than parsing scenarios

			// 1. Simple "new" with generic (should NOT become an <Element>)
			const a = new TrackedArray<number>();

			// 2. Multi-line generic with newline after '<'
			const b = new TrackedArray<
				string
			>();

			// // 3. Member expression + generic
			// class List<T> {
			// 	items: T[];
			// 	constructor() {
			// 		this.items = [];
			// 	}
			// }
			// class Containers {
			// 	List<T>() {
			// 		return new List<T>();
			// 	}
			// }

			// const c = new Containers.List<string>();

			// 4. Chained call with generic method
			const someSource = new Array<number>(1, 2, 3);
			const d = someSource.map<number>((x) => x * 2).filter<boolean>((x) => !!x);

			// 5. Nested generics
			const e = new Map<string, Promise<number>>();

			// // 6. Generic after a call expression result
			// function getBuilder<T>() {
			// 	return {
			// 		build<U>() {
			// 			return {
			// 				build<V>() {
			// 					return 42;
			// 				}
			// 			};
			// 		}
			// 	};
			// }
			// const f = getBuilder()<ResultType>().build<number>();

			// // 7. Generic following optional chaining
			// const maybe = {};
			// const g = maybe?.factory<number>()?.make<boolean>();

			// // 8. Comparison operator (ensure '<' here NOT misparsed as generics)
			// let x = 10, y = 20;
			// const h = x < y ? 'lt' : 'ge';

			// // 9. Chained comparisons with intervening generics
			// class Box<T> {
			// 	value: T;
			// 	constructor(value?: T) {
			// 		this.value = value;
			// 	}
			// 	open<U>() {
			// 		return new Box<U>();
			// 	}
			// }
			// const limit = 100;
			// const i = new Box<number>().value < limit ? 'ok' : 'no';

			// 10. JSX / Element should still work
			<div class="still-works">
				<span>{a.length}</span>
			</div>

			// // 11. Generic function call vs Element: Identifier followed by generic args
			// function identity<T>(value: T): T {
			// 	return value;
			// }
			// const j = identity<number>(42);

			// 12. Member + generic call immediately followed by another call
			class Factory {
				create<T>() {
					return (value: T) => value;
				}
			}
			const factory = new Factory();
			const k = factory.create<number>()(123);

			// // 13. Multiple generic segments in chain
			// function foo<T>() {
			// 	return {
			// 		bar<U>() {
			// 			return {
			// 				baz<V>() {
			// 					return true;
			// 				}
			// 			};
			// 		}
			// 	};
			// }
			// const l = foo<number>().bar<string>().baz<boolean>();

			// 14. Generic with constraint + default
			type Extractor<T extends { id: number } = { id: number }> = (v: T) => number;
			const m: Extractor = (v) => v.id;

			// // 15. Generic in angle after "new" + trailing call
			// class Wrapper<T> {
			// 	value: T;
			// 	constructor() {
			// 		this.value = null as unknown as T;
			// 	}
			// 	unwrap<U>() {
			// 		return null as unknown as U;
			// 	}
			// }
			// const n = new Wrapper<number>().unwrap<string>();

			// // 16. Angle brackets inside type assertion vs generic call
			// function getUnknown(): unknown {
			// 	return { a: 1 };
			// }
			// getUnknown.factory = function<T>() {
			// 	return {
			// 		make<U>() {
			// 			return 2;
			// 		}
			// 	};
			// };
			// const raw = getUnknown();
			// const o = (raw as Map<string, number>).get('a');

			// // 17. Generic with comma + trailing less-than comparison on next token
			// class Pair<T1, T2> {
			// 	first: T1;
			// 	second: T2;
			// 	constructor() {
			// 		this.first = null as unknown as T1;
			// 		this.second = null as unknown as T2;
			// 	}
			// }
			// const p = new Pair<number, string>();
			// const q = 1 < 2 ? p : null;

			// // 18. Nested generics with line breaks resembling JSX indentation
			// interface Node<T> {
			// 	value: T;
			// }
			// interface Edge<W> {
			// 	weight: W;
			// }
			// class Graph<N, E> {
			// 	nodes: N[];
			// 	edges: E[];
			// 	constructor() {
			// 		this.nodes = [];
			// 		this.edges = [];
			// 	}
			// }
			// const r = new Graph<
			// 	Node<string>,
			// 	Edge<number>
			// >();

			// // 19. Ternary containing generics in both branches
			// let flag = true;
			// const s = flag ? new Box<number>() : new Box<string>();

			// 20. Generic inside template expression
			const t = `length=${new TrackedArray<number>().length}`;

			// 21. Optional chaining + generic + property access
			const registry = new TrackedMap<string, number>();
			const u = registry.get<number>('id')?.toString();

			// // 22. Generic call used as callee for another call
			// function make<T>() {
			// 	return (value: T) => value;
			// }
			// const v = make<number>()('done');

			// // 23. Generic followed by tagged template (ensure not confused with JSX)
			// function tagFn<T>(strings: TemplateStringsArray, ...values: T[]) {
			// 	return values[0];
			// }
			// const tagResult = tagFn<number>`value`;

			// // 24. Sequence mixing: (a < b) + generic call in same statement
			// function compute<T>(x: T, y: T): T {
			// 	return y;
			// }

			// const w = (x < y) && compute<number>(x, y);



			// Additional component focusing on edge crankers

			// // 28. Generic after parenthesized new expression
			// const aa = (new Box<number>()).open<string>();

			// // 29. Generic chain right after closing paren of IIFE
			// class Builder<Kind> {
			// 	finalize<Result>() {
			// 		return {
			// 			result: null as unknown as Result
			// 		};
			// 	}
			// }
			// const builder = new Builder<Number>();
			// const result = (function(){ return builder; })()<Number>().finalize<boolean>();

			// // 30. Angle bracket start of conditional expression line
			// function adjust<T>(value: T): T {
			// 	return value;
			// }
			// const val =
			// 	new Wrapper<number>()
			// 		.value < 100
			// 		? adjust<number>(10)
			// 		: adjust<number>(20);


			// // 32. Generic with comments inside angle list
			// class Mapper<Key, Value> {
			// 	map: Map<Key, Value>;
			// 	constructor() {
			// 		this.map = new Map<Key, Value>();
			// 	}
			// }
			// const gg = new Mapper<
			// 	// key type
			// 	string,
			// 	/* value type */
			// 	number
			// >();

			// // 33. Map of generic instance as key
			// const mm = new Map<TrackedArray<number>, TrackedArray<string>>();

			// // 34. Inline assertion then generic call

			// const asserted = (getUnknown() as Factory).create<number>();
		}


		render(App);
	});

	it('supports rendering compositie components using <@component> syntax', () => {
		component App() {
			component basic() {
				<div>{'Basic Component'}</div>
			}

			const tracked_basic = track(() => basic);

			<@tracked_basic />
		}

		render(App);
		flushSync();

		expect(container.textContent).toBe('Basic Component');
	});

	it('supports rendering compositie components using <object.@component> syntax', () => {
		component App() {
			component basic() {
				<div>{'Basic Component'}</div>
			}

			const tracked_basic = track(() => basic);

			const obj = {
				tracked_basic,
			}

			<obj.@tracked_basic />
		}

		render(App);
		flushSync();

		expect(container.textContent).toBe('Basic Component');
	});

	it('supports rendering compositie components using <@object.@component> syntax', () => {
		component App() {
			component basic() {
				<div>{'Basic Component'}</div>
			}

			const tracked_basic = track(() => basic);

			const obj = {
				tracked_basic,
			}

			const tracked_object = track(obj);

			<@tracked_object.@tracked_basic />
		}

		render(App);
		flushSync();

		expect(container.textContent).toBe('Basic Component');
	});

	it('render simple text as children', () => {
		component App() {
			let name = 'Click Me';

			<Child
				class="my-button"
			>{name}</Child>;
		}

		component Child({children, ...rest}) {
			<button {...rest}><children /></button>
		}

		render(App);
		expect(container).toMatchSnapshot();
	});

	it('handles dynamic component switching', () => {
		component Child1() {
			<div>{"I am child 1"}</div>
		}

		component Child2() {
			<div>{"I am child 2"}</div>
		}

		component App() {
			let thing = track(() => Child1)

			<div id="container">
				<@thing />
			</div>

			<button onClick={() => @thing = @thing === Child1 ? Child2 : Child1}>{"Change Child"}</button>
		}

		render(App);
		flushSync();

		expect(container.querySelector('#container').textContent).toBe('I am child 1');

		const button = container.querySelector('button');
		button.click();
		flushSync();

		expect(container.querySelector('#container').textContent).toBe('I am child 2');

		button.click();
		flushSync();

		expect(container.querySelector('#container').textContent).toBe('I am child 1');
	});
});
