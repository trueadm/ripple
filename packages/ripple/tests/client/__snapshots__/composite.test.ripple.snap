// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`composite components > correct handles passing through component props and children 1`] = `
<div>
  <!---->
  <div>
    <div>
      I am A
    </div>
    <!---->
    <div>
      other text
    </div>
    <!---->
    <div>
      I am B
    </div>
    <!---->
  </div>
  <!---->
  
</div>
`;

exports[`composite components > render simple text as children 1`] = `
<div>
  <!---->
  <button
    class="my-button"
  >
    Click Me
    <!---->
  </button>
  <!---->
  
</div>
`;
